core/src/main/java/org/apache/iceberg/avro/SupportsIndexProjection.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.avro;
+
+import java.io.Serializable;
+import java.util.List;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.types.Types;
+
+public abstract class SupportsIndexProjection implements StructLike, Serializable {
+  private final int[] fromProjectionPos;
+
+  /** Noop constructor that does not project fields */
+  protected SupportsIndexProjection(int size) {
+    this.fromProjectionPos = new int[size];
+    for (int i = 0; i < fromProjectionPos.length; i++) {
+      fromProjectionPos[i] = i;
+    }
+  }
+
+  /** Base constructor for building the type mapping */
+  protected SupportsIndexProjection(Types.StructType baseType, Types.StructType projectionType) {
```

Comment:
I prefer direct imports for nested classes (if the context is clear) to shorten lines. Given that this is a new class, I'd consider importing `StructType` and `NestedField` directly. That said, our codebase isn't very consistent. Up to you.